gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,259 @@
+        pair_scores = np.zeros((self._num_images, self._num_images))
+        np.fill_diagonal(pair_scores, -np.inf)
+
+        # initialize empty lists to collect all possible depths for each view
+        depths: List[List[float]] = [[] for _ in range(self._num_images)]
+
+        for j in range(num_tracks):
+            track = self._sfm_result.get_track(j)
+            num_measurements = track.number_measurements()
+            measurements = [track.measurement(k) for k in range(num_measurements)]
+            w_x = track.point3()
+            for k1 in range(num_measurements):
+                for k2 in range(k1 + 1, num_measurements):
+                    i_a = measurements[k1][0]
+                    i_b = measurements[k2][0]
+
+                    key_a = -1
+                    key_b = -1
+
+                    # check if measurement j1 belongs to a valid camera a
+                    if i_a in self.camera_idx_to_patchmatchnet_idx:
+                        key_a = self.camera_idx_to_patchmatchnet_idx[i_a]
+                        # calculate track_i's depth in the camera pose
+                        a_z = self._sfm_result.get_camera(i_a).pose().transformTo(w_x)[-1]
+                        # update image i_a's depth list only when i_a in a valid camera
+                        depths[key_a].append(a_z)
+
+                    # check if measurement j2 belongs to a valid camera b
+                    if i_b in self.camera_idx_to_patchmatchnet_idx:
+                        key_b = self.camera_idx_to_patchmatchnet_idx[i_b]
+                        # calculate track_i's depth in the camera pose
+                        b_z = self._sfm_result.get_camera(i_b).pose().transformTo(w_x)[-1]
+                        # update image i_b's depth list only when i_b in a valid camera
+                        depths[key_b].append(b_z)
+
+                    # if both cameras are valid cameras
+                    if key_a >= 0 and key_b >= 0:
+                        # calculate score for track_i in the pair views (cam_a, cam_b)
+                        score_a_b = piecewise_gaussian(
+                            xPa=self._camera_centers[i_a] - w_x, xPb=self._camera_centers[i_b] - w_x
+                        )
+                        # sum up pair scores for each track_i
+                        pair_scores[key_a, key_b] += score_a_b
+                        pair_scores[key_b, key_a] += score_a_b
+
+        # sort pair scores, for i-th row, choose the largest (num_views-1) scores, the corresponding views are selected
+        #   as (num_views-1) source views for i-th reference view.
+        pairs = np.argsort(pair_scores, axis=1)[:, -self._num_views + 1 :][:, ::-1]
+
+        # filter out depth outliers and calculate depth ranges
+        depth_ranges = np.zeros((self._num_images, 2))
+        for i in range(self._num_images):
+            depths_sorted = sorted(depths[i])
+            depth_ranges[i, 0] = depths_sorted[int(len(depths[i]) * 0.01)]
```

Comment:
I think using np.percentile(depths_sorted, 0.01) would be clearer.
An inline comment explaining why these have to be cast to integers would be helpful too.

Reply:
0.01 and 0.99 (or whatever hyperparameters you choose) should be module level global constants like

MIN_DEPTH_PERCENTILE=0.02